packages/ozone/src/daemon/blob-diverter.ts

```diff
@@ -0,0 +1,206 @@
+import AtpAgent, { ComAtprotoSyncGetBlob } from '@atproto/api'
+import { SECOND } from '@atproto/common'
+import Database from '../db'
+import { retryHttp } from '../util'
+import { dbLogger } from '../logger'
+import { BlobReportServiceConfig } from '../config'
+
+type PollState = {
+  timer?: NodeJS.Timer
+  promise: Promise<void>
+}
+
+type Service = {
+  agent: AtpAgent
+  did: string
+}
+
+export class BlobDiverter {
+  destroyed = false
+
+  pollState: PollState = {
+    promise: Promise.resolve(),
+  }
+
+  appview: Service | undefined
+  pds: Service | undefined
+  serviceConfig: BlobReportServiceConfig
+
+  constructor(
+    public db: Database,
+    services: {
+      serviceConfig: BlobReportServiceConfig
+      appview?: {
+        url: string
+        did: string
+      }
+      pds?: {
+        url: string
+        did: string
+      }
+    },
+  ) {
+    this.serviceConfig = services.serviceConfig
+    if (services.appview) {
+      this.appview = {
+        agent: new AtpAgent({ service: services.appview.url }),
+        did: services.appview.did,
+      }
+    }
+    if (services.pds) {
+      this.pds = {
+        agent: new AtpAgent({ service: services.pds.url }),
+        did: services.pds.did,
+      }
+    }
+  }
+
+  start() {
+    this.poll(this.pollState, () => this.divertBlob())
+  }
+
+  poll(state: PollState, fn: () => Promise<void>) {
+    if (this.destroyed) return
+    state.promise = fn()
+      .catch((err) => {
+        dbLogger.error({ err }, 'blob divert failed')
+      })
+      .finally(() => {
+        state.timer = setTimeout(() => this.poll(state, fn), 30 * SECOND)
+      })
+  }
+
+  async processAll() {
+    await Promise.all([this.divertBlob(), this.pollState.promise])
+  }
+
+  async destroy() {
+    this.destroyed = true
+    const destroyState = (state: PollState) => {
+      if (state.timer) {
+        clearTimeout(state.timer)
+      }
+      return state.promise
+    }
+    await destroyState(this.pollState)
+  }
+
+  async divertBlob() {
+    const toPush = await this.db.db
+      .selectFrom('blob_divert_event')
+      .select('id')
+      .forUpdate()
+      .skipLocked()
+      .where('divertedAt', 'is', null)
+      .where('attempts', '<', 10)
+      .execute()
+    await Promise.all(toPush.map((evt) => this.attemptBlobDivert(evt.id)))
+  }
+
+  private async getBlob(opts: { did: string; cid: string }) {
+    // TODO: Is this safe to do or should we be reaching out to the pds instead?
```

Comment:
Need some help here figuring out the right way to get the blob.

Reply:
@devinivy might be able to fill in with typescript-specific tricks... maybe there is a helper for this already? You could look in the appview code, in the auto-labeler section.

The general flow is to resolve the DID doc (probably from a local cache, or even just passing the full doc through to this method from elsewhere); get the account's PDS from there; make an un-authenticated request to the `getBlobs` endpoint, with appropriate HTTP retries and timeout. For correctness, then verify the hash of the content bytes against the blob CID.